src/zkscript/merkle_trees/merkle_tree.py

```diff
@@ -0,0 +1,198 @@
+
+    Methods:
+        __init__(root: str, hash_function: str): Initializes the Merkle Tree. An assertion is
+            performed to ensure the hash function is valid and the root is hexadecimal.
+
+    """
+
+    def __init__(self, root: str, hash_function: str, depth: int):
+        assert all(c in "0123456789abcdefABCDEF" for c in root), f"{root} is not a valid hexadecimal string."
+
+        assert set(hash_function.split(" ")).issubset(
+            {"OP_RIPEMD160", "OP_SHA1", "OP_SHA256", "OP_HASH160", "OP_HASH256"}
+        ), f"{hash_function} is not a valid hash function."
+
+        assert depth > 0
+
+        self.root = root
+        self.hash_function = hash_function
+        self.depth = depth
+
+    def locking_merkle_proof_with_bit_flags(
+        self,
+        is_equal_verify: bool = False,
+    ) -> Script:
+        """Generate a locking script to verify a Merkle path.
+
+        The locking script returned by this function checks the validity of a Merkle path. For a root r, the
+        Merkle path for an element d is defined as a sequence:
+            (d, "", "") (h_1, aux_1, bit_1) ... (h_{self.depth-1}, aux_{self.depth-1}, bit_{self.depth-1})
+        So that:
+            - h_1 = sef.hash(d)
+            - h_i = self.hash(h_{i-1} || aux_{i-1}) if bit_{i-1} == 0 else self.hash(aux_{i-1} || h_{i-1})
+            - r = self.hash(h_{self.depth-1} || aux_{self.depth-1}) if bit_{self.depth-1} == 0
+                    else self.hash(aux_{self.depth-1} || h_{self.depth-1})
+
+        Args:
+            is_equal_verify: a boolean variable to choose between `OP_EQUAL` or `OP_EQUALVERIFY` as final opcode in the
+                script. Default to `False`.
+
+        Returns:
+            A locking script verifying a Merkle path.
+
+        Note:
+            - `self.hash_function` must be a combination of valid Bitcoin Script hash functions (e.g., `OP_SHA256`).
+            - `self.root` must be set to the expected Merkle root to verify against.
+            - the Merkle path is assumed to be passed as unlocking script in the following way:
+                stack = [h_{self.depth-1} aux_{self.depth-1} bit_{self.depth-1} ... aux_1 bit_1 d]
+
+        """
+
+        out = Script()
+
+        # stack in: [h_{self.depth-1} aux_{self.depth-1} bit_{self.depth-1} ... h_1 aux_1 bit_1 d]
+        # stack out: [<purported r>]
+        out += Script.parse_string(self.hash_function)
+        out += Script.parse_string(
+            " ".join([f"OP_SWAP OP_IF OP_SWAP OP_ENDIF OP_CAT {self.hash_function}"] * (self.depth - 1))
+        )
+
+        # stack in: [<purported r>]
+        # stack out: [fail if <purported r> != self.root else 1/""]
+        out.append_pushdata(bytes.fromhex(self.root))
+        out += Script.parse_string("OP_EQUALVERIFY") if is_equal_verify else Script.parse_string("OP_EQUAL")
+
+        return out
+
+    def locking_merkle_proof_with_two_aux(
+        self,
+        is_equal_verify: bool = False,
+    ) -> Script:
```

Comment:
The doc should explain what are aux_. It is not clear what is the difference with the previous function